### PR TITLE
implement _local_timestamps in DatetimeArrayMixin

### PR DESCRIPTION
**Pull request by jbrockmendel**
Broken out of #21715 because it is a blocker for a handful of orthogonal follow-ons.

### CODE REVIEW

File: pandas/core/indexes/datetimes.py [hunk 1]
```diff
@@ -603,14 +603,8 @@ def _local_timestamps(self):
...
         else:
-            values = self.asi8
-            indexer = values.argsort()
-            result = conversion.tz_convert(values.take(indexer), utc, self.tz)
-
-            n = len(indexer)
-            reverse = np.empty(n, dtype=np.int_)
-            reverse.put(indexer, np.arange(n))
-            return result.take(reverse)
+            # fall back to non-optimized implementation
```
Comment (jreback) [inline pandas/core/indexes/datetimes.py]: I would remove this comment

File: pandas/core/arrays/datetimes.py [hunk 2]
```diff
@@ -108,3 +109,16 @@ def _sub_datelike_dti(self, other):
             mask = (self._isnan) | (other._isnan)
             new_values[mask] = iNaT
         return new_values.view('timedelta64[ns]')
+
+    # -----------------------------------------------------------------
+    # Timezone Conversion and Localization Methods
+
+    def _local_timestamps(self):
+        values = self.asi8
```
Comment (jreback) [inline pandas/core/arrays/datetimes.py]: add a doc-string here (e.g. what are the return guarantees here)

**Review by jreback — CHANGES_REQUESTED** [review]
looks good, just some doc comments

File: pandas/core/arrays/datetimes.py [hunk 3]
```diff
@@ -108,3 +109,22 @@ def _sub_datelike_dti(self, other):
...
+    def _local_timestamps(self):
+        """
+        Convert to an i8 (unix-like nanosecond timestamp) representation
+        while keeping the local timezone and not using UTC.
+        This is used to calculate time-of-day information as if the timestamps
+        were timezone-naive.
+        """
+        values = self.asi8
+        indexer = values.argsort()
+        result = conversion.tz_convert(values.take(indexer), utc, self.tz)
```
Comment (jbrockmendel) [inline pandas/core/arrays/datetimes.py]: Does this only make sense if `tz_convert` requires its input to be sorted?  I don't think it does.

Comment (jreback) [reply pandas/core/arrays/datetimes.py]: yeah not sure why that indexer is there

Comment (jbrockmendel) [reply pandas/core/arrays/datetimes.py]: In the next pass I'll see if this can be simplified.

**Review by jreback — APPROVED** [review]

**Pull request merged**